### PR TITLE
[archive] Fix usage with TypeScript 4.8+

### LINKED ISSUE #6365: fix: typescript error TS2344 in fast-foundation.d.ts with TypeScript 4.8.2
# 🐛 Bug Report

Starting with TypeScript 4.8.2 there are errors like

```
node_modules/.pnpm/@microsoft+fast-foundation@2.46.14/node_modules/@microsoft/fast-foundation/dist/fast-foundation.d.ts:2252:66 - error TS2344: Type 'T' does not satisfy the constraint 'string | number | boolean | symbol | any[] | Uint8Array | ({ createCSS?(): string; } & Record<PropertyKey, any>) | null'.
  Type 'T' is not assignable to type '{ createCSS?(): string; } & Record<PropertyKey, any>'.
    Type 'T' is not assignable to type '{ createCSS?(): string; }'.

2252 declare function create<T>(nameOrConfig: string): CSSDesignToken<T>;

  node_modules/.pnpm/@microsoft+fast-foundation@2.46.14/node_modules/@microsoft/fast-foundation/dist/fast-foundation.d.ts:2252:25
    2252 declare function create<T>(nameOrConfig: string): CSSDesignToken<T>;
                                 ~
    This type parameter might need an `extends { createCSS?(): string; }` constraint.
  node_modules/.pnpm/@microsoft+fast-foundation@2.46.14/node_modules/@microsoft/fast-foundation/dist/fast-foundation.d.ts:2252:25
    2252 declare function create<T>(nameOrConfig: string): CSSDesignToken<T>;
                                 ~
    This type parameter might need an `extends { createCSS?(): string; } & Record<PropertyKey, any>` constraint.
  node_modules/.pnpm/@microsoft+fast-foundation@2.46.14/node_modules/@microsoft/fast-foundation/dist/fast-foundation.d.ts:2252:25
    2252 declare function create<T>(nameOrConfig: string): CSSDesignToken<T>;
                                 ~
    This type parameter might need an `extends string | number | boolean | symbol | any[] | Uint8Array | ({ createCSS?(): string; } & Record<PropertyKey, any>) | null` constraint.
```

`CSSDesignToken` is declared as
```ts
export declare interface CSSDesignToken<T extends string | number | boolean | BigInteger | null | Array<any> | symbol | ({
    createCSS?(): string;
} & Record<PropertyKey, any>)> extends DesignToken<T>, CSSDirective {
```

## 💻 Repro or Code Sample

[TS Playground Example of the error](https://www.typescriptlang.org/play?ts=4.8.2#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXzDigxAB4AVAPgApUoBbEAeRgGE9EsBzALngGcMMLKi4BKPqwDKUgCIh+3VORwBrEKgqUA3AChdIAB4AHHDAzxQkWAhEkYiKGATS5CpSvWby8IyVTA-AJCIlzwAD7wqMj0AEYgMBHwsTg4ECBQ+JEAQtwAkqgkXAlJ0RAQSQCCMDBQAJ6kmXWUSfx1cWlJ1ADeuvD9BEQkrgD81BLBwqJ6AL7wAGTwAErgZsCkAAowOMYJGHUA0iB1ADTwTZRilH4agfDyilzKahpa-b26MwYmZhZW0HB4HYEo5nPd3E9PK8fDcAkFBFMwpFonESpEUmkMll4LkuAUimioshylUavVGqhmq12ikKpFujMWr0ZkA)

## 🌍 Your Environment

TypeScript 4.8.2

### PR DESCRIPTION
**Pull request by fcollonval**
# Pull Request

## 📖 Description

Add typing to template variable to remove error seen in code consuming fast-foundation with TypeScript 4.8+

### 🎫 Issues

Fixes #6365 

### General

- [x] I have included a change request file using `$ yarn change`
- [ ] I have added tests for my changes.
- [ ] I have tested my changes.
- [ ] I have updated the project documentation to reflect my changes.
- [x] I have read the [CONTRIBUTING](https://github.com/microsoft/fast/blob/master/CONTRIBUTING.md) documentation and followed the [standards](/docs/community/code-of-conduct/#our-standards) for this project.

I'm not sure how to test for this. I force locally a TypeScript update to 4.9.5 and fixed the error mentioned in `@microsoft/fast-foundation`.

There are also error in `@microsoft/fast-router` but this does not fix them.

<details><summary>Errors in @microsoft/fast-router</summary>
<pre>
src/configuration.ts(49,9): error TS2322: Type 'DefaultRouteRecognizer<unknown>' is not assignable to type 'RouteRecognizer<TSettings>'.
  The types returned by 'recognize(...)' are incompatible between these types.
    Type 'Promise<RecognizedRoute<unknown> | null>' is not assignable to type 'Promise<RecognizedRoute<TSettings> | null>'.
      Type 'RecognizedRoute<unknown> | null' is not assignable to type 'RecognizedRoute<TSettings> | null'.
        Type 'RecognizedRoute<unknown>' is not assignable to type 'RecognizedRoute<TSettings>'.
          Type 'unknown' is not assignable to type 'TSettings'.
            'TSettings' could be instantiated with an arbitrary type which could be unrelated to 'unknown'.
src/recognizer.ts(271,17): error TS18047: 'segmentA' is possibly 'null'.
src/recognizer.ts(271,33): error TS18047: 'segmentB' is possibly 'null'.
src/recognizer.ts(275,17): error TS18047: 'segmentA' is possibly 'null'.
src/recognizer.ts(275,33): error TS18047: 'segmentB' is possibly 'null'.
src/router.ts(201,13): error TS2322: Type 'unknown' is not assignable to type 'Router<any> | null | undefined'.
</pre>
</details>

### CODE REVIEW

**Comment by chrisdholt** [comment]
Thanks for this @fcollonval - do we have a repro of this I can check out? I recall us seeing a similar TS issue which disappeared with a different fix which wasn't specific to 4.8. If I can validate the issue, happy to take an update/fix assuming it makes sense.

**Comment by chrisdholt** [comment]
> Thanks @chrisdholt
> 
> We have seen that first when working on https://github.com/brichet/jupyterlab/tree/toolbar_webcomponent (uses TypeScript 5.0.4).
> 
> But I bumped TypeScript (to 4.9.5) in the components library we are using built on fast v1 to ease error reproducibility: https://github.com/jupyterlab-contrib/jupyter-ui-toolkit/actions/runs/5600572055/jobs/10243153138?pr=63#step:9:200

Thanks for this, sorry for the delay. I'm going to fork and pull your repo to try and reproduce this. I have a feeling it may not be the exact TS issue. I also could be wrong :)

**Comment by fcollonval** [comment]
Thanks a lot for taking the time @chrisdholt

**Comment by m-akinc** [comment]
Any updates? Also interested in using TS 4.8+.

File: packages/web-components/fast-foundation/src/design-token/design-token.ts [hunk 1]
```diff
@@ -143,7 +143,7 @@ class DesignTokenImpl<T extends { createCSS?(): string }> extends CSSDirective
         return [...this._appliedTo];
     }
 
-    public static from<T>(
+    public static from<T extends { createCSS?(): string }>(
```
Comment (atmgrifter00) [inline packages/web-components/fast-foundation/src/design-token/design-token.ts:146]: Would it be good to create a couple of local types to consolidate some of these type concerns. Namely:

```ts
type DesignTokenType = string | number | boolean | BigInteger | null | Array<any> | symbol | {};

type CSSDesignTokenType = DesignTokenType |
        ({
              createCSS?(): string;
         } & Record<PropertyKey, any>);
```

Then you could do something like this everywhere:

```ts
public static from<T extends CSSDesignTokenType> ...
```

**Comment by rajsite** [comment]
The TypeScript version upgrade is blocking us from performing Angular upgrades without having all our applications disable library type checking.

[Angular 14 support ends in November](https://angular.io/guide/releases#actively-supported-versions) and we would like to be able to migrate to [Angular 16 which requires TypeScript >=4.9.3](https://angular.io/guide/versions#actively-supported-versions).

Ideally we can avoid falling out of supported Angular version usage so hopefully within the next week or two we would like FAST to be compatible with a newer TypeScript version so we can bubble up support through our libraries. @chrisdholt do you think that would be possible?

**Comment by chrisdholt** [comment]
> The TypeScript version upgrade is blocking us from performing Angular upgrades without having all our applications disable library type checking.
> 
> [Angular 14 support ends in November](https://angular.io/guide/releases#actively-supported-versions) and we would like to be able to migrate to [Angular 16 which requires TypeScript >=4.9.3](https://angular.io/guide/versions#actively-supported-versions).
> 
> Ideally we can avoid falling out of supported Angular version usage so hopefully within the next week or two we would like FAST to be compatible with a newer TypeScript version so we can bubble up support through our libraries. @chrisdholt do you think that would be possible?

We're focused on updating the latest alpha and beta versions, but it's requiring some changes to our testing libraries. The bulk of my effort for TS right now is going to focus on moving those versions forward towards stable and hopefully getting onto the latest TS version for alpha/beta versions so we can leverage TS's support for decorator metadata.

As it pertains to the archive branches - we can try to get some support there, but I would suggest starting to plan on how to move off those. Given the alpha/beta changes have been moving slower, we've continued to backport some fixes, but I don't see that going on forever. At some point, the code there is just difficult to maintain with backwards compatibility, especially given the amount of the time the FAST Components package has been deprecated. I'm keeping this open with the hopes we can get to updating the archive, but changes there are more difficult and time consuming as of now.

**Comment by chrisdholt** [comment]
> > The TypeScript version upgrade is blocking us from performing Angular upgrades without having all our applications disable library type checking.
> > [Angular 14 support ends in November](https://angular.io/guide/releases#actively-supported-versions) and we would like to be able to migrate to [Angular 16 which requires TypeScript >=4.9.3](https://angular.io/guide/versions#actively-supported-versions).
> > Ideally we can avoid falling out of supported Angular version usage so hopefully within the next week or two we would like FAST to be compatible with a newer TypeScript version so we can bubble up support through our libraries. @chrisdholt do you think that would be possible?
> 
> We're focused on updating the latest alpha and beta versions, but it's requiring some changes to our testing libraries. The bulk of my effort for TS right now is going to focus on moving those versions forward towards stable and hopefully getting onto the latest TS version for alpha/beta versions so we can leverage TS's support for decorator metadata.
> 
> As it pertains to the archive branches - we can try to get some support there, but I would suggest starting to plan on how to move off those. Given the alpha/beta changes have been moving slower, we've continued to backport some fixes, but I don't see that going on forever. At some point, the code there is just difficult to maintain with backwards compatibility, especially given the amount of the time the FAST Components package has been deprecated. I'm keeping this open with the hopes we can get to updating the archive, but changes there are more difficult and time consuming as of now.

Good news - I repro'd this locally when trying to update the repo. Let's see if @nicholasrice has any concerns with this; he may have ideas on where/why the constraint may be failing in this scenario as well.

**Comment by chrisdholt** [comment]
Update on this - I got some insight from the TS team on why this is throwing now. I'm going to push an update to TS which includes a fix for this. @nicholasrice as FYI.

**Comment by chrisdholt** [comment]
Made it through getting the repo building with TS 4.8+, unfortunately it requires changing out all the test suites due to ESM requirements. Building this locally and if it passes I plan to push it through. Thanks @fcollonval for this PR and all for your patience.

**Comment by fcollonval** [comment]
Thanks @chrisdholt for taking the time of looking at this maintenance PR.

**Comment by rajsite** [comment]
> Made it through getting the repo building with TS 4.8+, unfortunately it requires changing out all the test suites due to ESM requirements. Building this locally and if it passes I plan to push it through. Thanks @fcollonval for this PR and all for your patience.

That's awesome news, thanks @chrisdholt! Do you know when we might see the changes in a published package?

**Comment by rajsite** [comment]
> > Made it through getting the repo building with TS 4.8+, unfortunately it requires changing out all the test suites due to ESM requirements. Building this locally and if it passes I plan to push it through. Thanks @fcollonval for this PR and all for your patience.
> 
> That's awesome news, thanks @chrisdholt! Do you know when we might see the changes in a published package?

These changes are now in `@microsoft/fast-foundation@2.49.2` 🎉

**Pull request merged**